fixedwidth/src/main/java/org/apache/metamodel/fixedwidth/FixedWidthReader.java

```diff
@@ -31,156 +31,261 @@
+                line.append((char) ch);
+            }
+
+            if (ch == CARRIAGE_RETURN) {
+                readLineFeedIfFollows();
+            }
+
+            return (line.length()) > 0 ? line.toString() : null;
+        } else {
+            byte[] buffer = new byte[_expectedLineLength];
+            int bytesRead = _stream.read(buffer, 0, _expectedLineLength);
+
+            if (bytesRead < 0) {
+                return null;
+            }
+
+            return new String(buffer, _charsetName);
+        }
+    }
+
+    private void readLineFeedIfFollows() throws IOException {
+        _stream.mark(1);
+
+        if (_stream.read() != LINE_FEED) {
+            _stream.reset();
+        }
+    }
+
+    private boolean isEndingCharacter(int ch) {
+        return (ch == CARRIAGE_RETURN || ch == LINE_FEED || ch == END_OF_STREAM);
+    }
+
+    private boolean isEOLAvailable() {
+        return _eolPresent;
+    }
+
+    private void processCharacter(char c, StringBuilder nextValue, List<String> values, String recordData) {
+        nextValue.append(c);
+        final int valueWidth = getValueWidth(values, recordData);
+
+        if (nextValue.length() == valueWidth) {
+            addNewValueIfAppropriate(values, nextValue);
+            nextValue.setLength(0); // clear the buffer
+
+            if (_valueWidths != null) {
+                _valueIndex = (_valueIndex + 1) % _valueWidths.length;
+            }
+        }
+    }
+
+    private int getValueWidth(List<String> values, String recordData) {
+        if (_constantWidth) {
+            return _fixedValueWidth;
+        } else {
+            if (_valueIndex >= _valueWidths.length) {
+                if (_failOnInconsistentLineWidth) {
+                    String[] result = values.toArray(new String[values.size()]);
+                    throw new InconsistentValueWidthException(result, recordData, _rowNumber + 1);
+                } else {
+                    return -1; // silently ignore the inconsistency
+                }
+            }
+
+            return _valueWidths[_valueIndex];
+        }
+    }
+
+    private void addNewValueIfAppropriate(List<String> values, StringBuilder nextValue) {
+        if (_valueWidths != null) {
+            if (values.size() < _valueWidths.length) {
+                values.add(nextValue.toString().trim());
+            }
+        } else {
+            values.add(nextValue.toString().trim());
+        }
+    }
+
+    private String[] correctResult(String[] result) {
+        if (result.length != _valueWidths.length) {
+            String[] correctedResult = new String[_valueWidths.length];
+
+            for (int i = 0; i < result.length && i < _valueWidths.length; i++) {
+                correctedResult[i] = result[i];
+            }
+
+            result = correctedResult;
+        }
+
+        return result;
+    }
+
+    private InconsistentValueWidthException buildInconsistentValueWidthException(String recordData, String[] result,
```

Comment:
This method seems backwards to me. It is called "build[something]" but will usually (in the good cases) return null. And then it is evaluated later on to be thrown if it is not null. Why this pattern is needed I don't know. Seems a much more normal way to do it was to have something like a "validate[something]" which will throw the exception if it is desirable.